snow/engine/snowman/straggler_detect.go

```diff
@@ -0,0 +1,331 @@
+// See the file LICENSE for licensing terms.
+
+package snowman
+
+import (
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+	"golang.org/x/exp/maps"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/utils/set"
+
+	safemath "github.com/ava-labs/avalanchego/utils/math"
+)
+
+const (
+	minStragglerCheckInterval                   = time.Second
+	stakeThresholdForStragglerSuspicion         = 0.75
+	minimumStakeThresholdRequiredForNetworkInfo = 0.8
+	knownStakeThresholdRequiredForAnalysis      = 0.8
+)
+
+type stragglerDetectorConfig struct {
+	// getTime returns the current time
+	getTime func() time.Time
+
+	// minStragglerCheckInterval determines how frequently we are allowed to check if we are stragglers.
+	minStragglerCheckInterval time.Duration
+
+	// log logs events
+	log logging.Logger
+
+	// minConfirmationThreshold is the minimum stake percentage that below it, we do not check if we are stragglers.
+	minConfirmationThreshold float64
+
+	// connectedPercent returns the stake percentage of connected nodes.
+	connectedPercent func() float64
+
+	// connectedValidators returns a set of tuples of NodeID and corresponding weight.
+	connectedValidators func() set.Set[ids.NodeWeight]
+
+	// lastAcceptedByNodeID returns the last accepted height a node has reported, or false if it is unknown.
+	lastAcceptedByNodeID func(id ids.NodeID) (ids.ID, bool)
+
+	// processing returns whether this block ID is known and its descendants have not yet been accepted by consensus.
+	// This means that when the last accepted block is given as input, true is returned, as by definition
+	// its descendants have not been accepted by consensus, but this block is known.
+	// For any block ID belonging to an ancestor of the last accepted block, false is returned,
+	// as the last accepted block has been accepted by consensus.
+	processing func(id ids.ID) bool
+
+	// lastAccepted returns the last accepted block of this node.
+	lastAccepted func() ids.ID
+
+	// getSnapshot returns a snapshot of the network's nodes and their last accepted blocks,
+	// or false if it fails from some reason.
+	getSnapshot func() (snapshot, bool)
+
+	// haveWeFailedCatchingUp returns whether we have not replicated enough blocks of the given snapshot
+	haveWeFailedCatchingUp func(snapshot) bool
+}
+
+type stragglerDetector struct {
+	stragglerDetectorConfig
+
+	// continuousStragglingPeriod defines the time we have been straggling continuously.
+	continuousStragglingPeriod time.Duration
+
+	// previousStragglerCheckTime is the last time we checked whether
+	// our block height is behind the rest of the network
+	previousStragglerCheckTime time.Time
+
+	// prevSnapshot is the snapshot from a past iteration.
+	prevSnapshot snapshot
+}
+
+func newStragglerDetector(
+	getTime func() time.Time,
+	log logging.Logger,
+	minConfirmationThreshold float64,
+	lastAccepted func() (ids.ID, uint64),
```

Comment:
Why is an interface superior to a function pointer in this case? If I put an interface I would need to make a mock structure for it in the tests. Can you please explain?

Reply:
Generally interfaces provides more "defined" and flexible structures rather than pointer functions. I don't particularly like function pointers because they lock the function signature and all functions expecting the function pointer needs to be changed if it is modified. Also Mocking complex functions as a single interface can be much more easily done through mockgens. So in the end it's all readability, code standards and flexibility. IMHO having one or very specific pointer function is fine, but repeating it for the sake of easy testing is generally an issue.